File: solutions/490.the-maze.py
class Solution(object):
    def neighbors(self, node, maze):
        dirs = [[0,1],[0,-1],[1,0],[-1,0]]
        row = node[0]
        col = node[1]
        res = []
        for d in dirs:
            srow = row
            scol = col
            while srow >= 0 and srow < len(maze) and scol >= 0 and scol < len(maze[0]):                  
                if maze[srow][scol] == 1:
                    break
                    
                srow =  srow + d[0]
                scol = scol + d[1]
            res.append((srow-d[0], scol-d[1]))
            
        return res
        
    def hasPath(self, maze, start, destination):
        q = collections.deque()
        q.append((start[0], start[1]))
        
        visited = {}
        visited[(start[0],start[1])] = True
        
        while len(q) > 0:
            curr = q.pop()
            nn = self.neighbors(curr, maze)
            for n in nn:
                if n not in visited:
                    visited[n] = True
                    q.append(n)
            
        return (destination[0],destination[1]) in visited
            
            
            
        

File: solutions/490.the-maze.cpp
struct Pos {
    int row;
    int col;
};


class Solution {
public:
    vector<vector<int>> dirs = {{0,1}, {0,-1}, {1,0}, {-1, 0}};
    
    vector<Pos> getNeighbors(int srow, int scol, vector<vector<int>>& maze) {
        vector<Pos> poses;    
        
        for (vector<int> dir: dirs) {
            int row = srow + dir[0];
            int col = scol + dir[1];
                
            while (row >= 0 && row < maze.size() && col >= 0 && col < maze[0].size()) {
                if (maze[row][col] == 1) {
                    break;
                }
                
                row += dir[0];
                col += dir[1];
            }
            
            Pos pp;
            pp.row = row-dir[0];
            pp.col = col-dir[1];
            poses.push_back(pp);
        }
        
        return poses;
    }
    
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        vector<vector<bool>> visited(maze.size(), vector<bool>(maze[0].size(), false));
        
        queue<Pos> q;
        
        visited[start[0]][start[1]] = true;
        q.push({start[0], start[1]});
        
        while (!q.empty()) {
            Pos curr = q.front();
            q.pop();
            vector<Pos> neighbors = getNeighbors(curr.row, curr.col, maze);
            for (auto nn : neighbors) {
                if (!visited[nn.row][nn.col]) {
                    visited[nn.row][nn.col] = true;
                    q.push({nn.row,nn.col});
                }
            }
        }
        
        return visited[destination[0]][destination[1]];
        
    }
};